290 Provide UI for users to delete input + 125 Provide UI for Users to Rename The Input + 121 Implement functionality for users to Regenerate a new LinktaFlow

kinskiwu opened the pull request:
<!--# Pull Request Template -->
This PR contains 3 PRs:

## Description
**290 Provide UI for users to delete input**
- Implement the confirmation dialog with styling
- Error handling with SnackBarNotification is in place
- React Query setup in https://github.com/Linkta-org/core/issues/253 is integrated with the component

Fixes #290 

**125 Provide UI for Users to Rename The Input**
- Implemented RenameDialog for Users to rename user input
- Enable users to edit the title of UserInput through the UI.
- Display the new title in the UI as the user types.
- Implement input validation for input
- Error handling with SnackBarNotification is in place
- React Query setup in https://github.com/Linkta-org/core/issues/250 is integrated in the component
- Enhanced zod

Fixes #125 

**121 Implement functionality for users to Regenerate a new LinktaFlow**
- Renamed the useNewUserInputMutation hook to useCreateLinktaFlowMutation for better clarity
- Updated the hook to utilize the pre-configured axiosClient
- Added functionality for the regenerate button in the UserInputList component
- Integrated the useCreateLinktaFlowMutation hook to handle the regeneration process
- Mapped userinput _id to id
- Implemented SnackBar Notification

Fixes #121 

## Type of change
- [X] ✨  New feature (non-breaking change which adds functionality)

## How Can this be tested? Testing Plan to review this PR
- [X] Where in the app does this change take place?
- [X] What are the steps to test this change?

**290 Provide UI for users to delete input** 
1. At root dir, run `npm run dev`
2. Click a tab on the sidenav, such as 'tabby' -> open menu -> click 'Delete'
- [X] What is the expected output?
![Pasted Graphic 3](https://github.com/Linkta-org/core/assets/112212841/6fbf8cdb-c68c-43de-a144-f79fd111a9c7)
3. Click Delete -> redirected to `/output` & inputHistory on sidenav updated (tabby gets removed)
![Pasted Graphic 4](https://github.com/Linkta-org/core/assets/112212841/afa23cfe-6868-46e2-85c1-aa50b192bd4c)
4. Error msg when delete fails
![Failed to delete input  Please try again,](https://github.com/Linkta-org/core/assets/112212841/32b8bca7-944a-4a80-9e98-6d86846e616a)

**125 Provide UI for Users to Rename The Input**
1. At root dir, run npm run dev

Regular input:
Click 'tuxedo cats' -> Rename -> input 'new tuxedo cats' -> save
![image](https://github.com/Linkta-org/core/assets/112212841/a7e05dd7-c097-4ba6-bbfa-2e8e3f0b98bd)
![image](https://github.com/Linkta-org/core/assets/112212841/3f015dae-4863-4636-9829-d0f25bdeab27)
![image](https://github.com/Linkta-org/core/assets/112212841/e85b0b3b-c22e-423c-bdfb-88fd949d705e)

Error input:
Click 'tuxedo cats' -> Rename -> input <HTML> -> save
![image](https://github.com/Linkta-org/core/assets/112212841/1e7bdeb9-a59c-4b46-a629-bac879c60768)

- [X] What is the expected output?
Regular:
![image](https://github.com/Linkta-org/core/assets/112212841/9ae6bade-337d-46b5-958f-23d7c35f22b2)

Error:
![image](https://github.com/Linkta-org/core/assets/112212841/ad672e36-19d5-4bcd-94b2-3047c7b73456)

**121 Implement functionality for users to Regenerate a new LinktaFlow**
1. At root dir, run `npm run dev`
2. Selected one of the tabs on Sidenav and click `Regenerate`
![tamagotchi](https://github.com/Linkta-org/core/assets/112212841/7ba13224-073d-45e4-9584-5948309b90ff)
- [X] What is the expected output?
4. A new tab with the same title will appear on the top of the navbar stack
![one piece](https://github.com/Linkta-org/core/assets/112212841/ebd11def-190a-4447-a1ba-8e5d15559398)
5. You will be routed to a new output path with the new userInputId

6. When you encounter an error, you will see an error toast notification:
![image](https://github.com/Linkta-org/core/assets/112212841/6a8bc36b-5308-4484-86b5-4f211d2b29bc)

**Note:** If you see a 500 it means the llm response has an error, try a few times until you get the 200

## Checklist:

<!--Before submitting your pull request, please review the following checklist:-->

- [X] I have performed a self-review of my own code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [X] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [X] I have reviewed the [**Security Best Practices Document**](./docs/SECURITY_BEST_PRACTICES.md).
- [X] I have followed the [**Naming Conventions Guide**](./docs/NAMING_CONVENTIONS_GUIDE.md) for my changes.
- [X] I have ensured that my pull request title is descriptive.

kinskiwu committed:
implemented regenerate function

kinskiwu committed:
implemented rename ui

kinskiwu committed:
implemented delete ui

updated notification logic

removed incorrect mock linktaflow

updated mutate to mutateAsync

ejwoods commented on client/styles/layout/DeleteDialog.module.css:
The spacing between the `Delete` and `Cancel` buttons is slightly inconvenient UX - I think most sites tend to keep them side-by-side. I would recommend aligning this styling with the modal for renaming a dialog.

kinskiwu replied:
Thanks for the suggestion. Since this is a design topic, probably better to leave it for later when we review the design of the whole app together.

ejwoods commented on client/validators/userInputSchemas.ts:
```diff
@@ -19,5 +21,16 @@ const userInputValidationSchema = z.object({
+    .min(1, 'Input title is required')
+    .max(15, {
+      message: 'Title must be less than 15 characters in length!',
+    })
```
15 characters might be a little unnecessarily restrictive. Some of the existing prompts are over this - for example, 'teach a dog to sit' is 18 characters. It would probably be difficult to summarize complex topics in so few characters. I would align this number with the maximum input length when generating a flow.

kinskiwu replied:
Thanks for noting this. The restriction is for the input title - which is a temp solution at the moment and the 15 characters is to ensure the whole title can be viewed on the navbar. There is another ticket to enhance the title display.

ejwoods commented on client/components/layout/UserInputList.tsx:
```diff
@@ -7,74 +7,170 @@ import useDrawerStore from '@stores/userDrawerStore';
+          await queryClient.invalidateQueries({
+            queryKey: ['inputHistory'],
+          });
+          navigate('/output');
```
When navigating to `/output`, currently the NotFoundPage Placeholder is displayed, which makes sense since you need to be on a flow's page to delete it. Will something other than the 404 page be displayed here? 

Maybe we should navigate to '/generate' instead?

kinskiwu replied:
That's a good catch! I totally overlooked the 404. Thinking instead of the /output the /home-page may be more intuitive.

kinskiwu replied:
I agree /generate is a better destination. Just updated the path.

ejwoods reviewed (approved):
Re: 290, the deletion prompt looks great. Good use of the color palate to communicate the difference between the two buttons!

Re: 125, I'm seeing the input renamed. The modal looks great. 

Re: 121, the the Regenerate button works properly - a new tab is created and I can see differences in the output. If one hasn't already been created, I think there's potential for a ticket to add an integer to the new tabs created to allow users to differentiate between versions without needing to manually rename.

kinskiwu commented:
Just fyi, I updated the snackbar notification state management, removed the incorrect mock linktaflow file & created #312 for making a reusable Dialog component

LucieSeidler reviewed (changes requested):
Currently there's no rerouting if a user tries to go back to a deleted LinktaFlow, app continues to attempt to fetch LinktaFlow until another page is manually chosen

![fetchDeletedFlow](https://github.com/Linkta-org/core/assets/126367705/b540db99-90eb-43fe-a06f-3027c298b597)

kinskiwu commented:
@LucieSeidler I debugged the issue and created a hotfix to route to `/generate` when there is an error with the fetchLinktaFlow operation in `Flow`. Since this issue is related to overall error handling involving both the frontend and backend, I am going to create a separate ticket to enhance the error handling later tonight or tomorrow.
![image](https://github.com/Linkta-org/core/assets/112212841/6ab1f1c4-6c8b-45eb-9a05-72872f3f53c2)

colinvandergraaf reviewed (approved):
Overall, great work on this. Modal’s look awesome! Below are some questions/comments that came to mind when looking through this. No need to answer these here specifically here, bu for future discussion.  

290 - Delete Modal Input
Whats the difference between an Input and Linkta Flow Diagram? Aren’t these the same thing? 
 
125 Provide UI for Users to Rename
Why are special chars not allowed as title’s? What if I wanted to name a flow for an HTML element? 
I also noticed that it isn’t possible to submit the new title by clicking enter on my keyboard which is curious becaus the ubtton type is ‘submit’.  Do you have an idea why this might be? 

121  Regenerate Linkta Flow 
When clicking regenerate, I get a flicker of the background with the tree in between renders.   This may be worth investigating as part of a future ticket.  

Just out of curiosity, what are you doing to test the unhappy path to produce the errors in 290 and 121?  

Thanks for the note on the LLM response!

kinskiwu committed:
fixed submit keyboard issue

kinskiwu commented:
@colinvandergraaf 
Thanks for the review!

To answer your questions:
Q: Whats the difference between an Input and Linkta Flow Diagram? Aren’t these the same thing?
A: An Input refers to a UserInput which is the query a user puts in to generate a Linkta Flow Diagram. If you meant why we have separate userInput and linktaFlow documents instead of keeping the linktaFlow inside the userInput doc, it is because we may want to keep different versions of linktaFlows associated with the same userInput in a stretch feature, that's why we separated the lintaFlow from the userInput doc. 

Q: Why are special chars not allowed as title’s? What if I wanted to name a flow for an HTML element?
That's a good question and it was something I noticed too. I agree the current zod schemas for userInput could be improved. Since it needs further research, I created #321 to handle this matter. 

Q: I also noticed that it isn’t possible to submit the new title by clicking enter on my keyboard which is curious becaus the ubtton type is ‘submit’. Do you have an idea why this might be?
A: That's a good catch! I found that in order for the mui button to allow keyboard function, we need to wrap the button in a `form` component. Right now we are able to submit a title update form with the 'enter' key. Source: https://stackoverflow.com/questions/69766400/using-form-with-mui-onsubmit-method-not-working

Q: When clicking regenerate, I get a flicker of the background with the tree in between renders. This may be worth investigating as part of a future ticket.
A: Hmm I tried it but didn't notice the flicker. Maybe next time when we meet on zoom you can show us what the flickering looks like?

Q: Just out of curiosity, what are you doing to test the unhappy path to produce the errors in 290 and 121?
A: To trigger an error I manually threw errors in the related services on the API and the axios will get a 500 that triggers the snack bar.
